daemon/containerd/cache.go

```diff
@@ -2,211 +2,238 @@ package containerd
 
 import (
 	"context"
-	"reflect"
-	"strings"
+	"encoding/json"
+	"errors"
+	"fmt"
 
+	"github.com/containerd/containerd/content"
+	cerrdefs "github.com/containerd/containerd/errdefs"
 	"github.com/containerd/log"
 	"github.com/docker/docker/api/types/backend"
 	"github.com/docker/docker/api/types/container"
 	"github.com/docker/docker/builder"
 	"github.com/docker/docker/errdefs"
 	"github.com/docker/docker/image"
 	"github.com/docker/docker/image/cache"
+	"github.com/docker/docker/internal/multierror"
+	"github.com/docker/docker/layer"
 	"github.com/opencontainers/go-digest"
 	ocispec "github.com/opencontainers/image-spec/specs-go/v1"
 )
 
 // MakeImageCache creates a stateful image cache.
-func (i *ImageService) MakeImageCache(ctx context.Context, cacheFrom []string) (builder.ImageCache, error) {
-	images := []*image.Image{}
-	if len(cacheFrom) == 0 {
-		return &localCache{
-			imageService: i,
-		}, nil
+func (i *ImageService) MakeImageCache(ctx context.Context, sourceRefs []string) (builder.ImageCache, error) {
+	adaptor := cacheAdaptor{i}
+	if len(sourceRefs) == 0 {
+		return cache.NewLocal(adaptor), nil
 	}
 
-	for _, c := range cacheFrom {
-		h, err := i.ImageHistory(ctx, c)
+	cache := cache.New(adaptor)
+
+	for _, ref := range sourceRefs {
+		img, err := i.GetImage(ctx, ref, backend.GetImageOpts{})
 		if err != nil {
-			continue
-		}
-		for _, hi := range h {
-			if hi.ID != "<missing>" {
-				im, err := i.GetImage(ctx, hi.ID, backend.GetImageOpts{})
-				if err != nil {
-					return nil, err
-				}
-				images = append(images, im)
+			if errors.Is(err, context.Canceled) || errors.Is(err, context.DeadlineExceeded) {
+				return nil, err
 			}
+			log.G(ctx).Warnf("Could not look up %s for cache resolution, skipping: %+v", ref, err)
+			continue
 		}
+		cache.Populate(img)
 	}
 
-	return &imageCache{
-		lc: &localCache{
-			imageService: i,
-		},
-		images:       images,
-		imageService: i,
-	}, nil
+	return cache, nil
```

Comment:
This ends up super similar to the graphdrivers one. Wonder if we can refactor this further.

Reply:
Yeah it's exactly the same, just a different adaptor struct implementation.

Added a commit to extract the shared bits.

Reply:
Nice! It looks great now :)